### PR TITLE
Add example Update, Delete operation and PUT, DELETE endpoint in SQL db tutorial

### PR DESCRIPTION
**Pull request by akihiro-inui**
I have added an example code snippet for:
- Update, Delete operation for SQL database
- PUT, DELETE endpoint

### CODE REVIEW

**Comment by akihiro-inui** [comment]
Not exactly sure why the tests are failing

**Comment by Kludex** [comment]
@akihiro-inui `FAILED tests/test_tutorial/test_sql_databases/test_sql_databases.py::test_openapi_schema`

**Comment by Kludex** [comment]
Well, the code on the tutorials are tested as well, and you've performed some work there... 

`FAILED tests/test_tutorial/test_sql_databases/test_sql_databases.py::test_openapi_schema`

**Comment by akihiro-inui** [comment]
@Kludex That makes sense. Thanks!

**Comment by akihiro-inui** [comment]
Does anyone want to review?

File: docs/en/docs/tutorial/sql-databases.md [hunk 1]
```diff
@@ -417,6 +417,36 @@ The steps are:
 
     `Item(**item.dict(), owner_id=user_id)`
 
+
+### Update data
+Now create utility functions to update the existing data.
+The steps are: 
+
+* Find the data entry from the database and retrieve it as  SQLAlchemy model *instance*
```
Comment (ycd) [inline docs/en/docs/tutorial/sql-databases.md:425]: missing dot here.

File: docs/en/docs/tutorial/sql-databases.md [hunk 2]
```diff
@@ -417,6 +417,36 @@ The steps are:
 
     `Item(**item.dict(), owner_id=user_id)`
 
+
+### Update data
+Now create utility functions to update the existing data.
+The steps are: 
+
+* Find the data entry from the database and retrieve it as  SQLAlchemy model *instance* 
+* Replace the parameter in the model with the input value
```
Comment (ycd) [inline docs/en/docs/tutorial/sql-databases.md:426]: missing dot here.

File: docs/en/docs/tutorial/sql-databases.md [hunk 3]
```diff
@@ -417,6 +417,36 @@ The steps are:
...
+```Python hl_lines="27-33"
+{!../../../docs_src/sql_databases/sql_app/crud.py!}
+```
+
+
+### Delete data
+Now create utility functions to delete the existing data.
+The steps are: 
+
+* Find the data entry from the database and retrieve it as  SQLAlchemy model *instance*
```
Comment (ycd) [inline docs/en/docs/tutorial/sql-databases.md:440]: missing dot here.

File: docs_src/sql_databases/sql_app/crud.py [hunk 4]
```diff
@@ -24,6 +24,23 @@ def create_user(db: Session, user: schemas.UserCreate):
     return db_user
 
 
+def update_user(db: Session, user_id: int, new_password: str):
+    db_user_to_update = db.query(models.User).filter(models.User.id == user_id).first()
```
Comment (ycd) [inline docs_src/sql_databases/sql_app/crud.py:28]: @tiangolo uses `db_user` across all the documentation, i think you should stay with it.

File: docs_src/sql_databases/sql_app/crud.py [hunk 5]
```diff
@@ -24,6 +24,23 @@ def create_user(db: Session, user: schemas.UserCreate):
...
+    db_user_to_update.password = new_password
+    db.add(db_user_to_update)
+    db.commit()
+    db.refresh(db_user_to_update)
+    return db_user_to_update
+
+
+def delete_user(db: Session, user_id: int):
+    db_user_to_delete = db.query(models.User).filter(models.User.id == user_id).first()
```
Comment (ycd) [inline docs_src/sql_databases/sql_app/crud.py:37]:
```python
object.replace("_to_delete", "")
```
or :smile: 

```bash
sed -i 's/db_user_to_delete/db_user/g' docs_src/sql_databases/sql_app/crud.py

**Review by ycd — APPROVED** [review]
LGTM, just a few things.

**Comment by akihiro-inui** [comment]
Can we merge this?

**Comment by mikey-no** [comment]
Like the improved guide very much... I saw a small problem with the guide...
[CRUD Utils Introduction](https://6015cda713fd17d4463eef44--fastapi.netlify.app/tutorial/sql-databases/#crud-utils)
It says: "...although in this example we are only creating and reading."
Following your update this text will be incorrect.
Thanks
Mike

**Comment by akihiro-inui** [comment]
> Like the improved guide very much... I saw a small problem with the guide...
> [CRUD Utils Introduction](https://6015cda713fd17d4463eef44--fastapi.netlify.app/tutorial/sql-databases/#crud-utils)
> It says: "...although in this example we are only creating and reading."
> Following your update this text will be incorrect.
> Thanks
> Mike

It has been updated, Thanks for mentioning it :)

**Comment by mikey-no** [comment]
Hi akihiro-inui, 

I believe there is a typo in the update method in the most recent version of the documentation [Update data](https://601c4d88a74f51848969bbd6--fastapi.netlify.app/tutorial/sql-databases/#update-data).

The current line of code says:
`for var, value in vars(db_user).items():`

From my test of just this method in some toy example it did not work work. My fix was to change it to say this:
`for var, value in vars(user).items():`

The problem was the updates passed into the function were not read and applied to the existing record in the database.

I like not having to type out the attribute names of model being updated and that we are using a python function not pydantic based approach.

HTH

Thanks

Mike

File: docs_src/sql_databases/sql_app/crud.py [hunk 6]
```diff
@@ -24,6 +24,25 @@ def create_user(db: Session, user: schemas.UserCreate):
     return db_user
 
 
+def update_user(db: Session, user: schemas.UserUpdate):
+    db_user = db.query(user).filter(user.id == user.id).one()
+
+    # Update model class variable from requested fields
+    for var, value in vars(db_user).items():
```
Comment (akihiro-inui) [inline docs_src/sql_databases/sql_app/crud.py:31]:
```suggestion
    for var, value in vars(user).items():
```

**Review by akihiro-inui — COMMENTED** [review]
Fix following the suggestion from @mikey-no

**Comment by akihiro-inui** [comment]
Hi @mikey-no 

Thank you very much for suggesting the fix, it is indeed a typo.
I have updated the script accordingly.

File: docs_src/sql_databases/sql_app/crud.py [hunk 7]
```diff
@@ -24,6 +24,25 @@ def create_user(db: Session, user: schemas.UserCreate):
     return db_user
 
 
+def update_user(db: Session, user: schemas.UserUpdate):
+    db_user = db.query(user).filter(user.id == user.id).one()
```
Comment (akihiro-inui) [inline docs_src/sql_databases/sql_app/crud.py:28]:
```suggestion
    db_user = db.query(models.User).filter(user.id == user.id).one()
```

**Review by akihiro-inui — COMMENTED** [review]
Fix model user

**Comment by akihiro-inui** [comment]
The test is failing due to `ImportError: cannot import name 'RowProxy'` I consider this is not related to this PR.

**Comment by ArcLightSlavik** [comment]
```python
    db_user = crud.update_user(db, user_id=user_id, new_password=new_password)
```
```python
def update_user(db: Session, user: schemas.UserUpdate):
    db_user = db.query(models.User).filter(user.id == user.id).one()

    # Update model class variable from requested fields
    for var, value in vars(user).items():
        setattr(db_user, var, value) if value else None
    db.add(db_user)
    db.commit()
    db.refresh(db_user)
    return db_user
```
`TypeError: update_user() got an unexpected keyword argument 'user_id'`

**Comment by akihiro-inui** [comment]
> ```python
>     db_user = crud.update_user(db, user_id=user_id, new_password=new_password)
> ```
> 
> ```python
> def update_user(db: Session, user: schemas.UserUpdate):
>     db_user = db.query(models.User).filter(user.id == user.id).one()
> 
>     # Update model class variable from requested fields
>     for var, value in vars(user).items():
>         setattr(db_user, var, value) if value else None
>     db.add(db_user)
>     db.commit()
>     db.refresh(db_user)
>     return db_user
> ```
> 
> `TypeError: update_user() got an unexpected keyword argument 'user_id'`

This has been fixed. Thank you.

**Comment by Timusan** [comment]
How would you go about updating this with relationships? In the example above and in its tests the user is never updated with a new address. I would imagine you have a nested Pydantic model, just as in this example, but simply using 

```python
for var, value in vars(user).items():
    setattr(db_user, var, value) if value else None
```

Would throw an error complaining that the nested model would have no `_sa_instance_state` state.

**Comment by Netzvamp** [comment]
Found another issue with ``setattr(db_user, var, value) if value else None``. If you use boolean fields it's valid to use ``False`` as a value (e.g. is_active = False, is_public = False). I use ``setattr(db_user, var, value) if not None else None`` to fix that.

File: docs_src/sql_databases/sql_app/crud.py [hunk 8]
```diff
@@ -24,6 +24,25 @@ def create_user(db: Session, user: schemas.UserCreate):
     return db_user
 
 
+def update_user(db: Session, user: schemas.UserUpdate):
+    db_user = db.query(models.User).filter(user.id == user.id).one()
+
+    # Update model class variable from requested fields
+    for var, value in vars(user).items():
+        setattr(db_user, var, value) if value else None
```
Comment (cikay) [inline docs_src/sql_databases/sql_app/crud.py:32]: `value and setattr(db_user, var, value)` go ahead :)

Comment (akihiro-inui) [reply docs_src/sql_databases/sql_app/crud.py:32]: True, thank you. I will make this change and fix some failing test this weekend.

File: docs_src/sql_databases/sql_app/crud.py [hunk 9]
```diff
@@ -24,6 +24,25 @@ def create_user(db: Session, user: schemas.UserCreate):
     return db_user
 
 
+def update_user(db: Session, user: schemas.UserUpdate):
+    db_user = db.query(models.User).filter(user.id == user.id).one()
+
+    # Update model class variable from requested fields
+    for var, value in vars(user).items():
```
Comment (cikay) [inline docs_src/sql_databases/sql_app/crud.py:31]: Also `var` should be `key`. This is a convention.

**Review by cikay — CHANGES_REQUESTED** [review]

File: docs_src/sql_databases/sql_app/crud.py [hunk 10]
```diff
@@ -24,6 +24,25 @@ def create_user(db: Session, user: schemas.UserCreate):
     return db_user
 
 
+def update_user(db: Session, user: schemas.UserUpdate):
+    db_user = db.query(models.User).filter(user.id == user.id).one()
```
Comment (cikay) [inline docs_src/sql_databases/sql_app/crud.py:28]: Why comparision operator. Should not be single `=`

**Review by BilalAlpaslan — APPROVED** [review]

**Comment by tiangolo** [comment]
Thanks for the interest! :coffee: 

I wanted to re-write the SQL docs with SQLModel, I just finished that here: https://github.com/fastapi/fastapi/pull/12285

The new docs are here: https://fastapi.tiangolo.com/tutorial/sql-databases/

Given that, I'll pass on this one, but thanks for the effort! :cake:

**Pull request closed without merging**